src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+
+use {Point};
+
+
+/// A function to compute the inter-cluster distance between set `A` and `B`.
+pub type LinkageFunction = Fn(&Fn(usize, usize) -> f64, &[usize], &[usize]) -> f64;
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+#[derive(Copy, Clone)]
+pub enum LinkageCriterion<'a> {
+    /// Minimum or single-linkage clustering.
+    ///
+    /// `min { d(a, b) | a ∈ A, b ∈ B }`
+    Single,
+
+    /// Maximum or complete-linkage clustering.
+    ///
+    /// `max { d(a, b) | a ∈ A, b ∈ B }`
+    Complete,
+
+    /// An optimal version of complete-linkage clustering known as CLINK.
+    ///
+    /// **Beware:** It depends on the order of the elements and doesn't always return the best
+    /// dendrogram.
+    CLINK,
+
+    /// Custom likage criterion.
+    ///
+    /// By providing a custom `LinkageFunction` it's possible to define
+    /// a custom linkage criterion.
+    Custom(&'a LinkageFunction)
+}
+
+
+/// Agglomerative clustering, i.e., hierarchical bottum-up clustering.
+///
+/// Starting with each element in its own cluster, iteratively merge the two most similar
+/// clusters until a certain distance threshold is reached. The similarity between two
+/// clusters is mainly defined by the linkage criterion and the `Point`'s distance
+/// metric.
+///
+/// * **Single-linkage**: Implementation of the optimal SLINK [1] algorithm with O( n^2 ) time
+///   and O( n ) space complexity.
+/// * **Complete-linkage**: Currently a naive implementation with O( n^3 ) time
+///   and O( n^2 ) space complexity.
+/// * **CLINK**: Implementation of the optimal CLINK [2] algorithm with O( n^2 ) time
+///   and O( n ) space complexity.
+/// * **Custom**: _At least_ O( n^3 ) time and O( n^2 ) space complexity. The exact
+///   complexity depends on the linkage criterion.
+///
+/// # Examples
+///
+/// ```rust
+/// use cogset::{Agglomerative, Euclid, LinkageCriterion};
+///
+/// let data = [Euclid([0.0, 0.0]),
+///             Euclid([1.0, 0.5]),
+///             Euclid([0.2, 0.2]),
+///             Euclid([0.3, 0.8]),
+///             Euclid([0.0, 1.0])];
+///
+/// let agglomerative = Agglomerative::new(&data, LinkageCriterion::Single);
+///
+/// println!("Dendogram: {:#?}", agglomerative.dendrogram());
+/// ```
+///
+/// # References
+///
+/// [1]: Sibson, R. (1973). SLINK: an optimally efficient algorithm for the single-link
+///      cluster method. The Computer Journal, 16(1), 30-34.
+/// [2]: Defays, D. (1977). An efficient algorithm for a complete link method. The Computer
+///     Journal, 20(4), 364-366.
+#[derive(Clone, Debug)]
+pub struct Agglomerative<'a, P: 'a + Point> {
+    dendrogram: Box<Dendrogram<&'a P>>
+}
+
+impl<'a, P: 'a + Point> Agglomerative<'a, P> {
+    /// Computes the `Dendrogram` for the supplied `data` using `linkage` as criterion.
+    pub fn new<T>(data: T, linkage: LinkageCriterion) -> Agglomerative<'a, P>
+        where T: IntoIterator<Item=&'a P>
```

Comment:
Ah, ok, I thought a bit more, and made a patch that improves the situation: https://github.com/huonw/cogset/commit/a40812e4b116d42cc16ab1ff9ec5ccd82dbdf9d4

This means that this function could take `data: &[P]` and let people essentially do what this function does internally, i.e. if the caller has something other than a `Vec<P>` or `&[P]`, then instead of `Agglomerative::new(some_iter, ...)`, they can write 

```rust
let v = some_iter.into_iter().collect::<Vec<_>>(); // this is a Vec<&P>
Agglomerative::new(&v, ...)
```

Then again, it's not obvious to me how important it is to remove this allocation, since the actual dendrogram involves a lot of allocations itself. Thoughts?